File: backend/domain/scoring/sideways_consistency_score_calculator.go
package scoring

import (
	"pano_chart/backend/domain"
	"fmt"
	"math"
)

// SidewaysConsistencyScoreCalculator scores range-bound, oscillatory behavior.
type SidewaysConsistencyScoreCalculator struct{}

func (c *SidewaysConsistencyScoreCalculator) Name() string {
	return "Sideways Consistency"
}

func (c *SidewaysConsistencyScoreCalculator) Score(series domain.CandleSeries) (float64, error) {
	n := series.Len()
	if n < 6 {
		return 0, fmt.Errorf("at least 6 candles required")
	}
	closes := make([]float64, n)
	for i := 0; i < n; i++ {
		candle, _ := series.At(i)
		closes[i] = candle.Close()
	}
	// --- 1. Net Displacement Ratio (NDR) ---
	p0 := closes[0]
	pn := closes[n-1]
	minPrice, maxPrice := closes[0], closes[0]
	for _, v := range closes {
		if v < minPrice {
			minPrice = v
		}
		if v > maxPrice {
			maxPrice = v
		}
	}
	rangePrice := maxPrice - minPrice
	var ndr float64
	if rangePrice == 0 {
		// Flat line, per spec: score = 0
		return 0, nil
	} else {
		ndr = math.Abs(pn-p0) / rangePrice
	}
	if ndr < 0 {
		ndr = 0
	}
	if ndr > 1 {
		ndr = 1
	}
	// --- 2. Range Stability Score (RSS) ---
	window := 5
	if n < window {
		window = n
	}
	windowRanges := make([]float64, n-window+1)
	for i := 0; i <= n-window; i++ {
		wMin, wMax := closes[i], closes[i]
		for j := i; j < i+window; j++ {
			if closes[j] < wMin {
				wMin = closes[j]
			}
			if closes[j] > wMax {
				wMax = closes[j]
			}
		}
		windowRanges[i] = wMax - wMin
	}
	var sum, mean float64
	for _, v := range windowRanges {
		sum += v
	}
	mean = sum / float64(len(windowRanges))
	var stddev float64
	for _, v := range windowRanges {
		stddev += (v - mean) * (v - mean)
	}
	stddev = math.Sqrt(stddev / float64(len(windowRanges)))
	rss := 0.0
	if mean > 0 {
		rss = 1 - (stddev / mean)
	}
	if rss < 0 {
		rss = 0
	}
	if rss > 1 {
		rss = 1
	}
	// --- 3. Oscillation Density Score (ODS) ---
	extrema := 0
	for i := 1; i < n-1; i++ {
		if (closes[i] > closes[i-1] && closes[i] > closes[i+1]) || (closes[i] < closes[i-1] && closes[i] < closes[i+1]) {
			extrema++
		}
	}
	ods := 0.0
	if n > 2 {
		ods = float64(extrema) / float64(n-2)
	}
	if ods < 0 {
		ods = 0
	}
	if ods > 1 {
		ods = 1
	}
	// --- Final Score ---
	score := (1 - ndr) * rss * ods
	if score < 0 {
		score = 0
	}
	if score > 1 {
		score = 1
	}
	return score, nil
}

File: frontend/lib/domain/scoring/sideways_consistency_score_calculator.dart
import '../../domain/candle_series.dart';
import 'dart:math' as math;
import '../../features/candles/api/candle_response.dart';
import 'symbol_score_calculator.dart';

/// Sideways Consistency score: 1 - (max deviation from mean / mean)
class SidewaysConsistencyScoreCalculator implements SymbolScoreCalculator {
  static const int _rangeWindowSize = 5;

  @override
  double scoreSeries(CandleSeries series) {
    return score(CandleSeriesResponse(
        symbol: '', timeframe: '', candles: series.candles));
  }

  @override
  String get name => 'Sideways Consistency';

  @override
  double score(CandleSeriesResponse series) {
    final closes = series.candles.map((c) => c.close).toList();
    final n = closes.length;
    if (n < _rangeWindowSize + 1) {
      throw ArgumentError('At least ${_rangeWindowSize + 1} candles required');
    }
    // 1. Net Displacement Ratio (NDR)
    final p0 = closes.first;
    final pn = closes.last;
    final minPrice = closes.reduce(math.min);
    final maxPrice = closes.reduce(math.max);
    final range = maxPrice - minPrice;
    double ndr = 0.0;
    if (range > 0) {
      ndr = ((pn - p0).abs()) / range;
    }
    ndr = ndr.clamp(0.0, 1.0);

    // 2. Range Stability Score (RSS)
    final windowRanges = <double>[];
    for (int i = 0; i <= n - _rangeWindowSize; i++) {
      final window = closes.sublist(i, i + _rangeWindowSize);
      final wMin = window.reduce(math.min);
      final wMax = window.reduce(math.max);
      windowRanges.add(wMax - wMin);
    }
    final meanRange =
        windowRanges.reduce((a, b) => a + b) / windowRanges.length;
    double stddevRange = 0.0;
    if (windowRanges.length > 1) {
      final mean = meanRange;
      stddevRange = math.sqrt(windowRanges
              .map((r) => math.pow(r - mean, 2))
              .reduce((a, b) => a + b) /
          (windowRanges.length));
    }
    double rss = 1.0;
    if (meanRange > 0) {
      rss = 1 - (stddevRange / meanRange);
    }
    rss = rss.clamp(0.0, 1.0);

    // 3. Oscillation Density Score (ODS)
    int extremaCount = 0;
    for (int i = 1; i < closes.length - 1; i++) {
      if ((closes[i] > closes[i - 1] && closes[i] > closes[i + 1]) ||
          (closes[i] < closes[i - 1] && closes[i] < closes[i + 1])) {
        extremaCount++;
      }
    }
    double ods = 0.0;
    if (n > 2) {
      ods = extremaCount / (n - 2);
    }
    ods = ods.clamp(0.0, 1.0);

    // Final sideways score
    double sidewaysScore = (1 - ndr) * rss * ods;
    sidewaysScore = sidewaysScore.clamp(0.0, 1.0);
    return sidewaysScore;
  }

  @override
  String explain(CandleSeriesResponse series) {
    try {
      final s = score(series);
      if (s > 0.7) {
        return 'Highly consistent sideways action.';
      } else if (s > 0.2) {
        return 'Some sideways action.';
      } else {
        return 'Not consistent sideways.';
      }
    } catch (e) {
      return 'Not enough data to score.';
    }
  }
}
